### PR TITLE
MM-53023: Add origin device to ObserveAPIEndpointDuration

### PR DESCRIPTION
**Pull request by agarciamontoro**
#### Summary
This PR aims to open a discussion about the lack of origin device (web or mobile) in the API metrics. Having this data point in the metrics is useful not only for analyzing the origin device of the server's users, but also for isolating performance problems to the culprit frontend.

The final motivation of this has been the investigation I'm doing on the discrepancies between the new action frequencies in the load-test and the endpoint frequencies in Community: it would have been quite useful to isolate the origin of a specific endpoint whose trigger I'm trying to discover to a single codebase.

I'm open to discussing potential problems (as the reliability of the `IsMobileApp` function), as well as other alternative. This will also need some testing before merging, which I have not done, since I wanted to open this for feedback as early as possible.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53023

#### Screenshots
--

#### Release Note
```release-note
Added origin_client to the mattermost_api_time metrics
```

### CODE REVIEW

File: server/channels/web/handlers.go [hunk 1]
```diff
@@ -401,7 +401,15 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 				endpoint = h.HandlerName
 			}
-			c.App.Metrics().ObserveAPIEndpointDuration(endpoint, r.Method, statusCode, elapsed)
+			originDevice := "unknown"
+			if session := c.AppContext.Session(); session != nil {
+				originDevice = "web"
+				if session.IsMobileApp() {
+					originDevice = "mobile"
+				}
+			}
```
Comment (streamer45) [inline server/channels/web/handlers.go:410]: I wonder if we could rely on the request's user agent here to differentiate between the three platforms (web,desktop,mobile) when missing any better euristic. From what I remember mobile adds some special prefixes (e.g. `rn` and `Mattermost`). @cpoile Investigated some of this a while back, may be worth checking with him.

Desktop should also have a special `Mattermost` suffix.

Comment (cpoile) [reply server/channels/web/handlers.go:410]: Hey @agarciamontoro, like Claudio said, Desktop will always have a `Mattermost` in the user agent (also "electron", but `Mattermost` will be good enough). 
As for mobile, do you need to know if it's v2 or not?  Here's our full test for post-GA v2:

```go
func isMobilePostGA(r *http.Request) (mobile, postGA bool) {
	queryParam := r.URL.Query().Get("mobilev2")
	if queryParam == "true" {
		return true, true
	}

	// Below here is to test two things: is this mobile pre-GA? Is mobile version 441
	// (a one-week period when we didn't have the above queryParam)
	// TODO: simplify this once we can stop supporting 441.
	//   https://mattermost.atlassian.net/browse/MM-48929
	userAgent := r.Header.Get("User-Agent")
	fields := strings.Fields(userAgent)
	clientAgent := fields[0] // safe to assume there's at least one
	isMobile := strings.HasPrefix(clientAgent, "rnbeta") || strings.HasPrefix(clientAgent, "Mattermost")
	if !isMobile {
		return false, false
	}
	agent := strings.Split(clientAgent, "/")
	if len(agent) != 2 {
		return true, false
	}

	// We can't use a semver package, because we're not using semver correctly. So manually parse...
	version := strings.Split(agent[1], ".")
	if len(version) != 4 {
		return true, false
	}
	minor, err := strconv.Atoi(version[3])
	if err != nil {
		return true, false
	}
	return true, minor >= 441
}

```

So if it's any mobile version, checking for `Mattermost` prefix is good enough. Hope that helps?

Comment (agarciamontoro) [reply server/channels/web/handlers.go:410]: Oooh, this is good, thank you! I have one more question, though: you said that Desktop can be identified by having `Mattermost` in the user agent, but that's also what identifies mobile? I guess I'm missing something.

Comment (cpoile) [reply server/channels/web/handlers.go:410]: np: Desktop will have `Mattermost/version#` as the last field, while Mobile will have `Mattermost` as the first field.

Comment (agarciamontoro) [reply server/channels/web/handlers.go:410]: So intuitive! :joy: Thanks for sharing this obscure knowledge, I feel you must have suffered a bit to get it yourself.

Comment (cpoile) [reply server/channels/web/handlers.go:410]: Hah, just wait for when it gets changed. :)

**Comment by isacikgoz** [comment]
@agarciamontoro what is the status of this PR?

**Comment by agarciamontoro** [comment]
@isacikgoz, its status is "complicated". Now seriously, it's in my TODO list, but not forgotten. Will take care of this as soon as I have some spare cycles!

File: server/Makefile [hunk 2]
```diff
@@ -375,7 +375,7 @@ plugin-mocks: ## Creates mock files for plugins.
 
 einterfaces-mocks: ## Creates mock files for einterfaces.
 	$(GO) install github.com/vektra/mockery/v2/...@v2.23.2
-	$(GOBIN)/mockery --dir channels/einterfaces --all --output channels/einterfaces/mocks --note 'Regenerate this file using `make einterfaces-mocks`.'
+	$(GOBIN)/mockery --dir einterfaces --all --output einterfaces/mocks --note 'Regenerate this file using `make einterfaces-mocks`.'
```
Comment (agnivade) [inline server/Makefile:378]: Seems unrelated. Should we split to a different PR?

**Review by agnivade — APPROVED** [review]

File: server/Makefile (hunk 2 above)
Comment (agarciamontoro) [reply server/Makefile:378]: I needed this to run `make einterfaces-mocks` and recreate the mock for `ObserveAPIEndpointDuration`.

**Comment by agarciamontoro** [comment]
@cpoile I followed your suggestion (a simplified version of it, since I don't care about whether the mobile version is pre- or post-GA) in https://github.com/mattermost/mattermost/pull/23631/commits/892bce239d90ddeb5bfb34589a4ce860f797f131. I'd loved if you could take a look :)

File: server/channels/web/handlers.go [hunk 3]
```diff
@@ -406,11 +406,34 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 			} else {
 				endpoint = h.HandlerName
 			}
-			c.App.Metrics().ObserveAPIEndpointDuration(endpoint, r.Method, statusCode, elapsed)
+
+			originDevice := "web"
+			if isMobile(r) {
+				originDevice = "mobile"
+			}
```
Comment (streamer45) [inline server/channels/web/handlers.go:413]: I still wish we had a desktop differentiator as well. I understand the underlying client is pretty much the same as web but could still be helpful to understand where requests are coming from.

**Review by cpoile — APPROVED** [review]
The isMobile check looks good, thanks @agarciamontoro!

File: server/channels/web/handlers.go (hunk 3 above)
Comment (agarciamontoro) [reply server/channels/web/handlers.go:413]: Ah, yes, forgot about desktop at all :facepalm: Will do, thanks for the reminder!

**Comment by agarciamontoro** [comment]
Desktop identification added to the mix. Please review! :)

**Review by streamer45 — APPROVED** [review]
Thanks, approving with a couple of extra comments:

- Related comment I left on the other side https://github.com/mattermost/enterprise/pull/1465#discussion_r1307667003
- Unit test for `originDevice()` would carry some bonus points :p

**Comment by cwarnermm** [comment]
@agarciamontoro - Are there product or developer documentation updates needed for this PR?

**Comment by agarciamontoro** [comment]
@cwarnermm, we have documentation on the metrics: https://docs.mattermost.com/scale/performance-monitoring.html But we don't document the labels in each metric, which is the addition here. That would be good to do at some point, but not something we need to do now.

**Comment by cwarnermm** [comment]
@agarciamontoro - Are you open to creating a docs PR for these updates when you have bandwidth?

**Comment by agarciamontoro** [comment]
@cwarnermm, I'm not sure whether we want to do that manually. I'd prefer to have that documentation automatically created from the file where everything's defined: https://github.com/mattermost/enterprise/blob/master/metrics/metrics.go

**Comment by cwarnermm** [comment]
Now that's an excellent idea, @agarciamontoro! I'd prefer auto-generated docs too!

**Comment by agarciamontoro** [comment]
Do you know if there's any example of auto-generated docs coming from the `enterprise` repo? Or would that be a first?
